Feature/cli parsing (#36)
PPN-microbench/PPN-microbench: README.md

# PPN-microbench

A portable microbenchmarking tool for compute node caracterisation.

# Architecture support

This tool is developed and tested for **64 bit ARM and x86**

# Benchmarks

> Bellow is a list of currently implemented benchmarks, more info in the [wiki](https://github.com/ppn-microbench/ppn-microbench/wiki).

Reviewer: ```suggestion
Below is a list of currently implemented benchmarks; detailed information is available in the [wiki](https://github.com/ppn-microbench/ppn-microbench/wiki).
```

| Benchmark | Description |
|---|---|

> | CPU frequency | CPU clock frequency in single and multi-core loads |
> | CPU OPS | Max. operation per seconds for 32, 64 bit and vector integers and floats |
> | Core to core latency | Latency of core to core communications |
> | Cache latency | Latency of cache and ram access |

Reviewer: ```suggestion
| CPU frequency | CPU clock frequency in single- and multi-core light workloads |
| CPU OPS | Peak integer & floating-point operations per second for 32-, 64-bit and SIMD arithmetic workloads |
| Core-to-core latency | Latency of core-to-core communications |
| Cache latency | Latency of different cache level accesses |
```

# Running

## Via the run script

The simplest way to run the application is from the `run` script. Upon executing it, a pdf will be output to `report/out/` containing results for all benchmarks.

> The `run` script executes the following steps:
>  - building the application
>  - running the application
>  - creating a python vitrual environment if none exists
>  - creating the pdf report from the application output

Reviewer: ```suggestion
The `run` script executes the following steps:
1. Build the application
2. Run the application
3. Create a Python virtual environment if none exists
4. Generate a PDF report from the application output
 ```

> If you have an earlier version of python, you might need to install `python<version>-venv` through your package manager.

Reviewer: Which earlier version of Python though? What is the minimum required version?
```suggestion
If you have an earlier version of Python, you might need to install `python<version>-venv` through your package manager.
```

## Manually

### Building

In order to build the application, run the following commands:

```bash
# -G Ninja drastically improves build times
cmake -B build [-G Ninja]
cmake --build build/
```

The executable will be located in `build/main/`.

### Running

> You can now simply run the executable. Note that by default, the json will be output to `./`.

Reviewer: ```suggestion
You can now simply run the executable. Note that by default, the JSON will be dumped to `./`.
```

### Report generation

> To generate the pdf report, move into `report/`. The python expects to find the json results in `report/out.json`.

Reviewer: ```suggestion
To generate the PDF report, move into `report/`. The Python script expects to find the JSON results in `report/out.json`.
```

> `Mathplotlib` is the only required python package. `weasyprint` is also needed to convert an html file into the final pdf, but it is not strictly required. You can simply run `pip install -r requirements.txt` to install all dependencies.

Reviewer: ```suggestion
`mathplotlib` and `weasypring` are the only required Python packages. You can install both dependencies by running `pip install -r requirements.txt`.

> Finally, running the python is as easy as `python3 main.py`

Reviewer: ```suggestion
Finally, running the Python script is as easy as `python3 main.py`
```

## Arguments

> It is possible to run individual / groups of benchmarks by passing arguments to the script or application.

Reviewer: ```suggestion
It is possible to run individual/groups of benchmarks by passing arguments directly to the Python script or C++ application:
```
